### PR TITLE
Generic filtering  Implementation

### PR DESCRIPTION
**Pull request by monicasarbu**
The proposal for the generic filtering can be found here: https://github.com/elastic/beats/issues/451

In this implementation two actions are implemented:
 - include_fields
 - drop_fields

and both accept `fields` as argument:

<pre>
- include_fields:
    fields: ["proc.mem", "proc.cpu", "cmdline"]
</pre>

Each event MUST contain a few fields that cannot be removed. For now, these "mandatory" fields are: ["@timestamp", "beat", type", "count"]. If you configure one of these fields in the `drop_fields`, it will be ignored.

A field can be a dictionary or a value. Example: `proc.cpu.total_p", "proc", "proc.cpu".

__Special cases to consider:__

- [x] Ignore the fields that are not available in the event. Example:

<pre>
- include_fields: 
      fields: ["xx", "proc.mem", "proc.cpu"]
</pre>

- [x] If none of the include fields is part of the event, drop the event as it contains no information except the "mandatory" fields. In the previous example all the events that have `type != process` will be dropped as they don't contain any of the fields: `proc.mem`, `proc.cpu`.

- [x] Filter the event by applying multiple filter rules defined under `filter. We start with a copy of the event and apply the filter rules one by to the filtered event.

event -> filter rule 1 -> event 1 -> filter rule 2 -> event 2 ...

<pre>
- include_fields:
      fields: ["proc"]
- include_fields:
      fields: ["proc.cpu.total_p", "proc.mem.rss_p", "proc.cmdline"]
</pre>

- [x] The generic filtering implementation relies that the event contains only primitives and MapStr, so before calling the filter rules the event checker is called that verifies:
     - the event doesn't contain struct
     - the event doesn't contain pointer 

__Testing:__

- [x] if works with topbeat
- [x] if works with filebeat
- [x] if works with packetbeat

### CODE REVIEW

File: libbeat/common/mapstr.go [hunk 1]
```diff
@@ -39,6 +40,105 @@ func (m MapStr) Update(d MapStr) {
...
+func (m MapStr) Delete(key string) *MapStr {
+	keyParts := strings.Split(key, ".")
+	keysLen := len(keyParts)
+
+	mapp := m
+	for i := 0; i < keysLen-1; i++ {
+		keyPart := keyParts[i]
+
+		if _, ok := mapp[keyPart]; ok {
+			mapp = mapp[keyPart].(MapStr)
```
Comment (urso) [inline libbeat/common/mapstr.go]: panic possible here.

File: libbeat/common/mapstr.go [hunk 2]
```diff
@@ -39,6 +40,105 @@ func (m MapStr) Update(d MapStr) {
...
+	fromPointer := m
+	for i := 0; i < keysLen-1; i++ {
+		keyPart := keyParts[i]
+
+		if _, ok := fromPointer[keyPart]; ok {
+			if _, already := toPointer[keyPart]; !already {
+				toPointer[keyPart] = MapStr{}
+			}
+			fromPointer = fromPointer[keyPart].(MapStr)
+			toPointer = toPointer[keyPart].(MapStr)
```
Comment (urso) [inline libbeat/common/mapstr.go]: panic on cast failure

File: libbeat/common/mapstr.go [hunk 3]
```diff
@@ -39,6 +40,105 @@ func (m MapStr) Update(d MapStr) {
...
+	}
+	return &to
+}
+
+func (m MapStr) Clone() *MapStr {
+	result := MapStr{}
+
+	for k, v := range m {
+		result[k] = v
+	}
```
Comment (urso) [inline libbeat/common/mapstr.go]: consider recursive/deep copy as delete might remove values from shared sub-objects

File: libbeat/filter/filter.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	return filters.filters[index]
+}
+
+func (filters *FilterList) Filter(event *common.MapStr) *common.MapStr {
+
+	// clone the event at first, before starting filtering
+	filtered := event.Clone()
+
+	for _, filter := range filters.filters {
+		filtered = filter.Filter(filtered)
```
Comment (urso) [inline libbeat/filter/filter.go]: what if result is nil?

File: libbeat/filter/filter.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	Fields []string `yaml:"fields"`
+}
+
+type FilterConfig struct {
+	DropFields    *DropFieldsConfig    `yaml:"drop_fields"`
+	IncludeFields *IncludeFieldsConfig `yaml:"include_fields"`
+}
+
+type FilterRule interface {
+	Filter(event *common.MapStr) *common.MapStr
```
Comment (urso) [inline libbeat/filter/filter.go]: why use pointers? common.MapStr is already passed by reference

File: libbeat/publisher/client.go [hunk 6]
```diff
@@ -78,32 +79,49 @@ func newClient(pub *PublisherType) *client {
...
 
 func (c *client) PublishEvents(events []common.MapStr, opts ...ClientOption) bool {
+	publishEvents := []common.MapStr{}
+
 	for _, event := range events {
-		c.annotateEvent(event)
+		publishEvent := c.alterEvent(event)
+		if publishEvent != nil {
+			publishEvents = append(publishEvents, *publishEvent)
+		}
```
Comment (urso) [inline libbeat/publisher/client.go]: loop ok. Go idiomatic In-place copy trick (no additional allocations):

    publishEvents := events[:0]
    ...

If publishEvents is empty no need to forward events + trigger Signal within context to indicate success (filtered events should be treated as send successfully).

Doing filtering early plays very well with 'Guaranteed' flag.

File: libbeat/publisher/client.go [hunk 7]
```diff
@@ -78,32 +79,49 @@ func newClient(pub *PublisherType) *client {
 }
 
 func (c *client) PublishEvent(event common.MapStr, opts ...ClientOption) bool {
-	c.annotateEvent(event)
-
+	publishEvent := c.alterEvent(event)
+	if publishEvent == nil {
+		return false
```
Comment (urso) [inline libbeat/publisher/client.go]: should return true to indicate success. Beat only cares about sending, but should not be affected filtering. Also signal success to Signal callback in ctx.

File: libbeat/common/mapstr.go (hunk 3 above)
Comment (monicasarbu) [inline libbeat/common/mapstr.go]: Good point! I'll fix it.

File: libbeat/filter/filter.go (hunk 5 above)
Comment (monicasarbu) [inline libbeat/filter/filter.go]: Agree that should not pass pointer. I wanted to do that anyway. But do you refer that needs to return no pointer as well? It's important to return a pointer as Filter might change the pointer as well.

Comment (urso) [inline libbeat/filter/filter.go]: but filter normally just returns another common.MapStr. Due to MapStr just being a typedef for map no pointer is required. We can even return a tuple `(newEvent common.MapStr, drop bool)` to make intention more clear.

Comment (monicasarbu) [inline libbeat/filter/filter.go]: I like the idea of returning drop, so I already implemented it.

File: libbeat/beat/beat.go [hunk 8]
```diff
@@ -213,10 +216,19 @@ func (b *Beat) LoadConfig() error {
 		return fmt.Errorf("error Initialising publisher: %v\n", err)
 	}
 
+	filters, err := filter.New(b.Config.Filter)
+	if err != nil {
+		fmt.Printf("Error Initialising filters: %v\n", err)
+		logp.Critical(err.Error())
```
Comment (andrewkroh) [inline libbeat/beat/beat.go]: I noticed we spell "initializing" two different ways in messages in this file. We should stop choose one way and change them all.

File: libbeat/common/mapstr.go [hunk 9]
```diff
@@ -39,6 +40,119 @@ func (m MapStr) Update(d MapStr) {
...
+
+	if _, ok := fromPointer[keyParts[keysLen-1]]; ok {
+		toPointer[keyParts[keysLen-1]] = fromPointer[keyParts[keysLen-1]]
+	} else {
+		return nil
+	}
+	return nil
+}
+
+func (m MapStr) Clone() MapStr {
```
Comment (monicasarbu) [inline libbeat/common/mapstr.go]: Maybe replace it with copystructure?

**Comment by monicasarbu** [comment]
Closing this PR in favor of https://github.com/elastic/beats/pull/1120.

**Pull request closed without merging**